onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -311,34 +313,45 @@ def destroy(self, request, *args, **kwargs):
 
                 initial_count = self.object.submission_count()
                 queryset = Instance.objects.filter(
-                    id__in=instance_ids,
-                    xform=self.object,
-                    # do not update this timestamp when the record have
-                    # already been deleted.
-                    deleted_at__isnull=True
-                )
-                # loop through queryset
-                # then call delete_instance that calls .save()
-                # to allow emitting post_save signal
-                for instance in queryset_iterator(queryset):
-                    delete_instance(instance, request.user)
-                # updates the num_of_submissions for the form.
-                after_count = self.object.submission_count(force_update=True)
-                number_of_records_deleted = initial_count - after_count
-
-                # send message
-                send_message(
-                    instance_id=instance_ids, target_id=self.object.id,
-                    target_type=XFORM, user=request.user,
-                    message_verb=SUBMISSION_DELETED)
-
-                return Response(
-                    data={
-                        "message":
-                        "%d records were deleted" % number_of_records_deleted
-                    },
-                    status=status.HTTP_200_OK
-                )
+                        id__in=instance_ids,
+                        xform=self.object,
+                        # do not update this timestamp when the record have
+                        # already been deleted.
+                        deleted_at__isnull=True
+                    )
+                # Use Async when deleting subsets of submissions > 100
+                # Or when deleting all submissions and len(instance_ids) > 100
+                if delete_all_submissions and len(instance_ids) > 100 or (
+                    len(instance_ids)
+                        > settings.DELETE_SUBMISSION_ASYNC_THRESHOLD):
+                    delete_subissions_async(
+                        instances=queryset,
```

Comment:
Should this be `delete_submissions_async.delay()`

Reply:
Also wondering if celery will be able to parse a queryset

Reply:
Should we also give a way for users to track the progress of the deletes or a response that states the delete will happen asynchronous or something of the sort ?